### PR TITLE
quic: support loopback for QUIC connections

### PR DESCRIPTION
**Pull request by mmcgee-jump**
We are trying to send outgoing packets for 127.0.0.1 to the external facing NIC rather than on the loopback device, which won't work.

This is mostly to support development and testing, although it's conceivable someone would send transactions via. loopback in production as well.

### CODE REVIEW

File: src/app/fdctl/run/tiles/fd_net.c [hunk 1]
```diff
@@ -197,6 +219,12 @@ during_frag( void * _ctx,
   fd_memcpy( ctx->frame, src, sz ); // TODO: Change xsk_aio interface to eliminate this copy
 }
 
+typedef struct __attribute__((packed)) {
+  fd_eth_hdr_t eth[1];
+  fd_ip4_hdr_t ip4[1];
+  fd_udp_hdr_t udp[1];
+} eth_ip_udp_t;
+
```
Comment (ptaffet-jump) [inline src/app/fdctl/run/tiles/fd_net.c:227]: This was just debugging code and isn't needed anymore

File: src/app/fdctl/run/tiles/fd_net.c [hunk 2]
```diff
@@ -6,6 +6,10 @@
 #include "../../../../tango/xdp/fd_xdp.h"
 #include "../../../../tango/xdp/fd_xsk_private.h"
 
+#include "../../../../util/net/fd_eth.h"
+#include "../../../../util/net/fd_ip4.h"
+#include "../../../../util/net/fd_udp.h"
```
Comment (ptaffet-jump) [inline src/app/fdctl/run/tiles/fd_net.c:11]: Also unnecessary I think

File: src/app/fdctl/Local.mk [hunk 3]
```diff
@@ -54,6 +54,9 @@ fdctl: $(OBJDIR)/bin/fdctl
 $(OBJDIR)/bin/solana: solana/target/$(RUST_PROFILE)/solana
 	$(MKDIR) $(dir $@) && cp solana/target/$(RUST_PROFILE)/solana $@
 
+$(OBJDIR)/bin/solana-keygen: solana/target/$(RUST_PROFILE)/solana-keygen
```
Comment (mmcgee-jump) [inline src/app/fdctl/Local.mk:57]: Prefer to use `fdctl keygen` maybe add a faucet keygen for it?

File: src/app/fdctl/configure/sysctl.c [hunk 4]
```diff
@@ -21,6 +21,8 @@ static const char * params[] = {
   "/proc/sys/net/core/bpf_jit_enable",
   "/proc/sys/fs/file-max",
   "/proc/sys/fs/nr_open",
+  "/proc/sys/net/ipv4/conf/lo/rp_filter",
```
Comment (mmcgee-jump) [inline src/app/fdctl/configure/sysctl.c:24]: Document under `initializing.md`

File: src/app/fdctl/run/tiles/fd_net.c [hunk 5]
```diff
@@ -243,29 +233,8 @@ after_frag( void *             _ctx,
...
     if( FD_UNLIKELY( *opt_sz < sizeof( eth_ip_udp_t ) ) )
-      FD_LOG_ERR(( "loopback packet too small %lu", *opt_sz ));
-
-    fd_memset( hdr->eth->dst, 0, 6UL );
-    fd_memset( hdr->ip4->daddr_c, 0, 4UL );
-    fd_memset( hdr->ip4->saddr_c, 0, 4UL );
-    hdr->ip4->check = 0U;
-    /* TODO: Do we need to update check here? */
-    hdr->ip4->check = fd_ip4_hdr_check( ( fd_ip4_hdr_t const *) FD_ADDRESS_OF_PACKED_MEMBER( hdr->ip4 ) );
+      FD_LOG_WARNING(( "loopback packet too small %lu", *opt_sz ));
```
Comment (mmcgee-jump) [inline src/app/fdctl/run/tiles/fd_net.c:237]: Why change this to WARNING? Should never happen.

File: src/app/fdctl/Local.mk (hunk 3 above)
Comment (llamb-jump) [reply src/app/fdctl/Local.mk]: I added the subcommand `fddev keygen pubkey <path-to-file>` which is used in the new test.

File: src/app/fdctl/run/tiles/fd_net.c (hunk 5 above)
Comment (llamb-jump) [reply src/app/fdctl/run/tiles/fd_net.c]: I ended up removing this check - I had changed it to warning so we wouldn't crash at runtime if sending a small packet due to a bug. LMK if you want it added back.

File: src/app/fdctl/keygen.c [hunk 6]
```diff
@@ -84,6 +94,36 @@ generate_keypair( const char * keyfile,
   if( FD_UNLIKELY( setegid( gid ) ) ) FD_LOG_ERR(( "setegid() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
 }
 
+void
+keygen_pubkey( const char * file_path ) {
+  FILE * fp = fopen( file_path, "r" );
```
Comment (mmcgee-jump) [inline src/app/fdctl/keygen.c:99]: You can use `load_key_into_protected_memory( path, 1 )` here instead.

File: src/app/fdctl/keygen.c [hunk 7]
```diff
@@ -7,18 +7,28 @@
...
+  KEYGEN_KEY_VOTE_ACCOUNT,
+  KEYGEN_PUBKEY,
+} keygen_key_type_t;
 
 void
 keygen_cmd_args( int *    pargc,
                  char *** pargv,
                  args_t * args) {
-  char * usage = "usage: keygen <identity|vote>";
+  char * usage = "usage: keygen <identity|vote|pubkey>";
```
[the remark this answers is not in the document]
Comment (llamb-jump) [reply src/app/fdctl/keygen.c:20]: This matches the solana invocation `solana-keygen pubkey <file>`.

File: .github/workflows/functional_tests.yml [hunk 8]
```diff
@@ -6,14 +6,14 @@ on:
...
-  single-transaction:
+  functional-tests:
     runs-on:
       group: github-v1
     env:
       CC: gcc
     steps:
       - name: Install ausearch for debugging seccomp/bad syscalls
-        run: sudo apt update && sudo apt install -y auditd
+        run: sudo apt update && sudo apt install -y auditd libudev-dev
```
Comment (mmcgee-jump) [inline .github/workflows/functional_tests.yml:16]: Think this dep can be removed now?

File: src/app/fdctl/keygen.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mmcgee-jump) [reply src/app/fdctl/keygen.c:20]: `solana-keygen` is their top level binary, so they have subcommands underneath that (`new`, `pubkey`, `grind`, ...). We don't have that, fdctl is the top level, so subcommands mostly would go under there.

If you want to have a `keys` category or something which mirrors the structure of `solana-keygen` I don't mind that, but then the args would be `<new|pubkey>` and new would take a subarg of `<identity|vote>`.

[the remark this answers is not in the document]
Comment (mmcgee-jump) [reply src/app/fdctl/keygen.c:20]: (This is a public interface that will be frozen for all of time once it has a few external consumers so I think it's worth thinking about this even if it's incidental to the PR)

[the remark this answers is not in the document]
Comment (mmcgee-jump) [reply src/app/fdctl/keygen.c:20]: It's fine to do that, but it should be hierarchical. Do you disagree about the hierarchy?

[the remark this answers is not in the document]
Comment (llamb-jump) [reply src/app/fdctl/keygen.c]: I updated it to be:
```
usage:
  keys new identity
  keys new vote
  keys pubkey <path-to-keyfile>
  ```

File: .github/workflows/functional_tests.yml (hunk 8 above)
Comment (mmcgee-jump) [reply .github/workflows/functional_tests.yml:16]: But you're adding it as part of this change? What requires it? I think it was here when you built the `solana-keygen` command but isn't needed now?

Comment (llamb-jump) [reply .github/workflows/functional_tests.yml]: You're right good catch, updated.

**Review by ptaffet-jump — APPROVED** [review]
Great!

**Pull request merged**